### ISSUE TITLE
useFeatureFlag data should be available at page load to avoid flash

### ISSUE THREAD

**Issue by sqs** [sourcegraph/sourcegraph#45849]
If a feature flag (and `useFeatureFlag`) is used to show/hide a global navbar item or anything else visible on the initial page load, there is an additional delay or flash after the rest of the UI loads before the feature flag boolean is available. This makes feature flags less usable than user settings `experimentalFeatures` because the via (and `getExperimentalFeatures(settingsCascade.final)`) is available on the first render cycle for GlobalNavbar.

I see that user settings are also not synchronously provided on the initial page load (they are also loaded async), but we trigger the user settings fetch well before we trigger the feature flags fetch. So either we should do the same with feature flags, or maybe we should move both to being provided in the HTML sent from the server so they are truly available on the initial page load.

(Not sure who to tag on this. @vovakulikov I see you were possible working on something similar ("background feature flags evaluation")?)

**Comment by vovakulikov** [comment]
@sqs I agree with this statement about async loading feature flags. We don't have so many actual feature flags that it would be a problem to include them in the initial load (have them in the `window.context`, maybe?) 

I'm currently not working on the feature flags. I think @erzhtor has more context about feature flag evaluation. 

I had similar ideas about license flags, though (move them in the `window.context` on the backend, this should be fast because our license checks are just reading the configuration license token and extracting info about sub-products (code insights, batch-changes ...) license keys.  This should be fast enough to deprecate the async license check and, by this reduce the numbers of loading states when we open code-insights, batch-changes, or <any other page behind license check> https://github.com/sourcegraph/sourcegraph/issues/43718

**Comment by erzhtor** [comment]
We request feature flags asynchronously because:
1. they are evaluated on the backend
2. they are evaluated on demand, meaning that the user is assigned a specific feature flag value only if they have experienced a feature behind that feature flag. Its required to reliably determine results of feature flag based AB test/experiments.

Two options to make feature flags synchronous (and thus avoid flash):
1. Ideal solution: feature flag evaluation should be moved to the client side (similar to launch darkly and other feature flagging systems). In short, on the initial load, the client will load information about feature flags and their metadata, once a certain feature flag is being used, the client will evaluate synchronously in the browser/frontend and reply with the feature flag value, and asynchronously send it to the backend for usage in event logs/metrics.
2. Quick solution: evaluate all feature flags on initial page load. This easier to achieve, however it will be not a proper feature flag assignment /evaluation as user might not even experience a feature behind certain feature flag but still be assigned to it. As a result we cannon reliably determine in event logs/ Amplitude and thus say if AB test/experiment was successful or not.
